666 repair model attribute typing and docstrings

sallymatson opened the pull request:
# Description

Move attribute declarations out of `_setup` to subclass `__init__` functions. The attributes are now defined and documented in the proper place (e.g. see [updated hydrology docs](https://virtual-ecosystem.readthedocs.io/en/666-repair-model-attribute-typing-and-docstrings/api/models/hydrology/hydrology_model.html#module-virtual_ecosystem.models.hydrology.hydrology_model) as compared to [old hydrology docs](https://virtual-ecosystem.readthedocs.io/en/latest/api/models/hydrology/hydrology_model.html#module-virtual_ecosystem.models.hydrology.hydrology_model)).

It would be great to get general feedback, as well as everyone looking over the changes to their modules in order to make sure the variables are all typed correctly etc @vgro @davidorme @jacobcook1995 @TaranRallings. 

Fixes #666 

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works
- [ ] Relevant documentation reviewed and updated

sallymatson committed:
Moved class attribute definitions to init

sallymatson committed:
Removed the function to prevent basemodel subclass init

sallymatson committed:
Fixing sphinx error

dalonsoa reviewed (approved):
The changes look good, and indeed this is a more canonical approach than using fake class variables.

Probably the docs should be updated to explain what goes in `__init__` and what in `_setup` when creating a new model.

davidorme reviewed (changes requested):
This is looking good. Can you update the doc API files for the models to include inherited members? I think that's what we want as we currently can't see anything being inherited from the base class.

For example:

https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/7838fa8fa6dfa884748cd330464d27ae58e781f8/docs/source/api/models/plants/plants_model.md?plain=1#L29-L36

should go to (I think):

````md
```{eval-rst}
.. automodule:: virtual_ecosystem.models.plants.plants_model
    :autosummary:
    :members:
    :exclude-members: model_name
    :inherited_members:
```
````

sallymatson committed:
Adding inherited members to API dpcs.

sallymatson committed:
Add clarification about init vs setup in base model.

sallymatson commented:
> This is looking good. Can you update the doc API files for the models to include inherited members? I think that's what we want as we currently can't see anything being inherited from the base class.
> 
> For example:
> 
> https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/7838fa8fa6dfa884748cd330464d27ae58e781f8/docs/source/api/models/plants/plants_model.md?plain=1#L29-L36
> 
> should go to (I think):
> 
> ```
> ```{eval-rst}
> .. automodule:: virtual_ecosystem.models.plants.plants_model
>     :autosummary:
>     :members:
>     :exclude-members: model_name
>     :inherited_members:
> ```
> ```

I added this. I'm a bit confused about how the docs work TBH because in the version that I built and linked in the description above, it does seem to be showing? Is that in a different place? Anyways, added `:inherited-members:` to all of them !

davidorme reviewed (approved):
This looks good - we really need to update the model creation docs, but we also need to do that to reflect the static model changes as well, so lets do that separately under #676

davidorme committed:
Merge branch 'develop' into 666-repair-model-attribute-typing-and-docstrings
